test/server/config_validation/xds_verifier.cc

```diff
@@ -0,0 +1,325 @@
+#include "test/server/config_validation/xds_verifier.h"
+
+#include "common/common/logger.h"
+
+namespace Envoy {
+
+XdsVerifier::XdsVerifier(test::server::config_validation::Config::SotwOrDelta sotw_or_delta)
+    : num_warming_(0), num_active_(0), num_draining_(0), num_added_(0), num_modified_(0),
+      num_removed_(0) {
+  if (sotw_or_delta == test::server::config_validation::Config::SOTW) {
+    sotw_or_delta_ = SOTW;
+  } else {
+    sotw_or_delta_ = DELTA;
+  }
+  ENVOY_LOG_MISC(debug, "sotw_or_delta_ = {}", sotw_or_delta_);
+}
+
+/**
+ * get the route referenced by a listener
+ */
+std::string XdsVerifier::getRoute(const envoy::config::listener::v3::Listener& listener) {
+  envoy::config::listener::v3::Filter filter0 = listener.filter_chains()[0].filters()[0];
+  envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager conn_man;
+  filter0.typed_config().UnpackTo(&conn_man);
+  return conn_man.rds().route_config_name();
+}
+
+/**
+ * @return true iff the route listener refers to is in all_routes_
+ */
+bool XdsVerifier::hasRoute(const envoy::config::listener::v3::Listener& listener) {
+  return all_routes_.contains(getRoute(listener));
+}
+
+bool XdsVerifier::hasActiveRoute(const envoy::config::listener::v3::Listener& listener) {
+  return active_routes_.contains(getRoute(listener));
+}
+
+/**
+ * prints the currently stored listeners and their states
+ */
+void XdsVerifier::dumpState() {
+  ENVOY_LOG_MISC(debug, "Listener Dump:");
+  for (const auto& rep : listeners_) {
+    ENVOY_LOG_MISC(debug, "Name: {}, Route {}, State: {}", rep.listener.name(),
+                   getRoute(rep.listener), rep.state);
+  }
+}
+
+/*
+ * if you add a new listener that is not being updated, there are two cases:
+ * 1. the listener has not been added before
+ *    - if envoy knows about the listener's route, it will be added as active, else warming
+ * 1. a listener of the same name has been added before but it now draining
+ *    - if the routes match, the old listener will be added back to active
+ *    - if the routes don't match, the listener will be added as active/warming as relevant
+ *
+ * if a listener is updated (i.e. there is a already a listener by this name), there are 3 cases:
+ * 1. the old listener is active and the new is warming:
+ *    - old will remain active
+ *    - new will be added as warming, to replace the old when it gets its route
+ * 2. the old listener is active and new is active:
+ *    - old's filter chain is drained (seemingly instantaneously)
```

Comment:
What do you mean by filter chain drain here? Listeners own filter chain factories, the actual filter chain belongs to an active connection, which doesn't just disappear when a listener is updated.

Reply:
I thought I read something about filter chains draining somewhere but not able to find it now, I must have been mistaken.  I think it should just be that the old listener is drained instead.  Still not too sure why this listener drains instantaneously and does not show up as draining in config dumps while if you explicitly remove a listener (even one with no active connections), it shows up as draining.